text/0000-structural-records.md

```diff
@@ -0,0 +1,1262 @@
+`serde`, if there is a will, to offer implementations of `Serialize` for all
+structural records.
+
+Note that while `impl serde::Serialize for T { ... }` may not be possible
+without extensions, the following would be:
+
+```rust
+#[derive(Serialize)]
+struct RectangleTidy {
+    dimensions: {
+        width: u64,
+        height: u64,
+    },
+    color: {
+        red: u8,
+        green: u8,
+        blue: u8,
+    },
+}
+```
+
+## "Auto-implementing traits is a magical hack"
+
+Indeed, we would much prefer to use a less magical approach,
+but providing these traits without compiler magic would require
+significantly more complexity such as polymorphism over field names
+as well as variadic generics. Therefore, to make structural records
+usable in practice, providing a small set of traits with compiler
+magic is a considerably less complex approach. In the future, perhaps
+the magic could be removed.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The syntax we've used is chosen to be maximally consistent
+with positional tuples and with nominally typed braced structs.
+In particular, `Foo(x, y)` for tuple structs becomes `(x, y)`
+for the structural variant, `Foo { a: x, b: y }` becomes
+`{ a: x, b: y }`, `(x,)` becomes `{ a: x }`, and the pattern
+`(x, y, ..)` becomes `{ a: x, b: y, .. }`.
+
+We could require the prefix `struct` as in `struct { a: x }`.
+This would do away with the backtracking and also permit
+one-field structural records without having a disambiguating
+comma in the tail. However, this would be both less ergonomic
+as well as less consistent with the rest of the language.
+
+As for not doing this, this feature is not so critical that
+the absence of it would be a crisis, but it does smooth out
+the language, provide more consistency, and better readability
+when compared with the structurally typed positional tuples
+that we already have.
+
+## We once removed structural records, why re-add them?
+
+[@pcwalton]: https://internals.rust-lang.org/t/why-were-structural-records-removed/1553/2
+
+In 2015, [@pcwalton] noted that we once removed structural records from the
+language because:
+
+> They had three usability problems:
+>
+> 1. The fields had to be in the exact same order every time you constructed an
+> instance of the record. This is because the compiler needed a canonical ordering.
+>
+> 2. They could not be recursive without the compiler doing fancy typechecking
+> (coinductive IIRC), which was felt to be not worth the effort.
+>
+> 3. Field-level privacy can’t work with them.
+
+Point 1. does not apply to this proposal. You can use any order you like when
+writing out the type, an expression, or pattern, as long as all the fields are
+there and of the same type once the fields have been sorted lexicographically. 
+
+Point 2. and 3. still hold, but this is and has always been true of positional
+tuples so it is nothing new.
```

Comment:
So for all structural records both the type and the fields are always public basically? That's how tuples do.

Reply:
Yup, that's correct; I'll try to clarify that.